diff --git a/ai_provider_amazeeio.module b/ai_provider_amazeeio.module
index b82fef8031566e9d1e53e658102f2e0f5e349c82..efbe0879f2bbe479bf1dd1dc4745d4e2200ba6ae 100644
--- a/ai_provider_amazeeio.module
+++ b/ai_provider_amazeeio.module
@@ -5,6 +5,7 @@
  * Hook implementations for the Amazee.io AI provider.
  */

+use Drupal\Core\StringTranslation\TranslatableMarkup;
 use Drupal\ai_search\Plugin\search_api\backend\SearchApiAiSearchBackend;
 use Drupal\Core\Session\AccountInterface;
 use Drupal\user\UserInterface;
@@ -82,3 +83,22 @@ function _ai_provider_amazeeio_amazee_logout(): void {
     \Drupal::logger('ai_provider_amazeeio')->error('Failed to log out of Amazee.io: @error', ['@error' => $e->getMessage()]);
   }
 }
+
+/**
+ * Implements hook_menu_links_discovered_alter().
+ *
+ * Alter descriptions of Open-AI provider menu links to not confuse users.
+ */
+function ai_provider_amazeeio_menu_links_discovered_alter(&$links) {
+  $openai = $links['ai_provider_openai.settings_menu'];
+
+  $links['ai_provider_litellm.settings_menu']['description'] = _ai_provider_amazeeio_inactive_description($links['ai_provider_litellm.settings_menu']['description']);
+  $links['ai_provider_openai.settings_menu']['description'] = _ai_provider_amazeeio_inactive_description($links['ai_provider_openai.settings_menu']['description']);
+}
+
+/**
+ * Adjust the description of inactive menu items.
+ */
+function _ai_provider_amazeeio_inactive_description(TranslatableMarkup $input) {
+  return new TranslatableMarkup('@description<br/><em class="form-item__description"><strong>amazee.ai</strong> builds on this provider, but it is not used unless you explicitly configure it.</em>', ['@description' => $input]);
+}
diff --git a/ai_provider_amazeeio.services.yml b/ai_provider_amazeeio.services.yml
index 19c375944bd770f3f786cc4ab79744671a2f96ec..2cb816233118f10028da74df453f79b75ee49e89 100644
--- a/ai_provider_amazeeio.services.yml
+++ b/ai_provider_amazeeio.services.yml
@@ -1,9 +1,4 @@
 services:
-  ai_provider_amazeeio.route_subscriber:
-    class: Drupal\ai_provider_amazeeio\Routing\RouteSubscriber
-    tags:
-      - { name: event_subscriber }
-
   ai_provider_amazeeio.api_client:
     class: Drupal\ai_provider_amazeeio\AmazeeIoApi\AmazeeClient
     arguments: ['@http_client', '@logger.factory']
diff --git a/src/Plugin/VdbProvider/AmazeeioVdbProvider.php b/src/Plugin/VdbProvider/AmazeeioVdbProvider.php
index e5d51a2d6a9762ce335d7da5174a04101bd61f3a..04e63ff34472782d366414d23096e85a143af721 100644
--- a/src/Plugin/VdbProvider/AmazeeioVdbProvider.php
+++ b/src/Plugin/VdbProvider/AmazeeioVdbProvider.php
@@ -74,6 +74,7 @@ class AmazeeioVdbProvider extends PostgresProvider {
    */
   public function getConnectionData() {
     $config = $this->getConfig();
+    $output = [];
     $output['host'] = $this->configuration['host'] ?? $config->get(key: 'postgres_host');
     // Fail if host is not set.
     if (!$output['host']) {
@@ -86,7 +87,10 @@ class AmazeeioVdbProvider extends PostgresProvider {
     $token = $config->get(key: 'postgres_password');
     $output['password'] = '';
     if ($token) {
-      $output['password'] = $this->keyRepository->getKey(key_id: $token)->getKeyValue();
+      $key = $this->keyRepository->getKey(key_id: $token);
+      if ($key) {
+        $output['password'] = $key->getKeyValue();
+      }
     }
     if (!empty($this->configuration['password'])) {
       $output['password'] = $this->configuration['password'];
diff --git a/src/Routing/RouteSubscriber.php b/src/Routing/RouteSubscriber.php
deleted file mode 100644
index 204f372ee444dcfa986cef2fbf128affff216677..0000000000000000000000000000000000000000
--- a/src/Routing/RouteSubscriber.php
+++ /dev/null
@@ -1,28 +0,0 @@
-<?php
-
-namespace Drupal\ai_provider_amazeeio\Routing;
-
-use Drupal\Core\Routing\RouteSubscriberBase;
-use Symfony\Component\Routing\RouteCollection;
-
-/**
- * A route subscriber to remove settings forms for providers this builds upon.
- */
-class RouteSubscriber extends RouteSubscriberBase {
-
-  /**
-   * {@inheritdoc}
-   */
-  protected function alterRoutes(RouteCollection $collection) {
-    if ($route = $collection->get('ai_provider_openai.settings_form')) {
-      $route->setRequirement('_access', 'FALSE');
-    }
-    if ($route = $collection->get('ai_provider_litellm.settings_form')) {
-      $route->setRequirement('_access', 'FALSE');
-    }
-    if ($route = $collection->get('ai_vdb_provider_postgres.settings_form')) {
-      $route->setRequirement('_access', 'FALSE');
-    }
-  }
-
-}

