diff --git a/modules/ai_provider_amazeeio_test/ai_provider_amazeeio_test.info.yml b/modules/ai_provider_amazeeio_test/ai_provider_amazeeio_test.info.yml
new file mode 100644
index 0000000000000000000000000000000000000000..cf67fc860f314617ee9fdcbd156e5f90685e3411
--- /dev/null
+++ b/modules/ai_provider_amazeeio_test/ai_provider_amazeeio_test.info.yml
@@ -0,0 +1,7 @@
+type: module
+name: AI Provider amazee.io testing
+description: Mock amazee.io services for testing purposes.
+package: Testing
+hidden: true
+dependencies:
+  - ai_provider_amazeeio:ai_provider_amazeeio
diff --git a/modules/ai_provider_amazeeio_test/ai_provider_amazeeio_test.services.yml b/modules/ai_provider_amazeeio_test/ai_provider_amazeeio_test.services.yml
new file mode 100644
index 0000000000000000000000000000000000000000..d84047b9e9a7d8f422f8bf6260679d5a8008dae7
--- /dev/null
+++ b/modules/ai_provider_amazeeio_test/ai_provider_amazeeio_test.services.yml
@@ -0,0 +1,7 @@
+services:
+  ai_provider_amazeeio_test.http_client:
+    class: Drupal\ai_provider_amazeeio_test\MockHttpClient
+    public: false
+    decorates: http_client
+    decoration_priority: 3
+    arguments: ['@ai_provider_amazeeio_test.http_client.inner']
diff --git a/modules/ai_provider_amazeeio_test/src/MockHttpClient.php b/modules/ai_provider_amazeeio_test/src/MockHttpClient.php
new file mode 100644
index 0000000000000000000000000000000000000000..67ab4d28e531dd5f02ae0050272f27e9ec31ebf8
--- /dev/null
+++ b/modules/ai_provider_amazeeio_test/src/MockHttpClient.php
@@ -0,0 +1,286 @@
+<?php
+
+namespace Drupal\ai_provider_amazeeio_test;
+
+use Drupal\ai_provider_amazeeio\Form\AmazeeioAiConfigForm;
+use GuzzleHttp\Client;
+use GuzzleHttp\Exception\ClientException;
+use GuzzleHttp\Psr7\Request;
+use GuzzleHttp\Psr7\Response;
+use GuzzleHttp\Utils;
+use Psr\Http\Message\ResponseInterface;
+use Symfony\Component\HttpFoundation\ParameterBag;
+
+/**
+ * Mock the http_client service.
+ *
+ * @phpstan-ignore class.extendsFinalByPhpDoc
+ */
+class MockHttpClient extends Client {
+
+
+  /**
+   * The decorated http_client service.
+   *
+   * @var \GuzzleHttp\Client
+   */
+  protected Client $innerService;
+
+  public function __construct(Client $inner_service) {
+    $this->innerService = $inner_service;
+  }
+
+  /**
+   * Mocked endpoints map.
+   *
+   * @value List of mocked endpoints, mapping to methods that execute them.
+   */
+  protected array $requests = [
+    'POST:/auth/validate-email' => 'mockValidateEmail',
+    'POST:/auth/sign-in' => 'mockSignIn',
+    'GET:/auth/me' => 'mockMe',
+    'GET:/regions' => 'mockRegions',
+    'POST:/private-ai-keys' => 'mockPostPrivateKey',
+    'GET:/private-ai-keys' => 'mockGetPrivateKeys',
+    'GET:/ch1/key/info' => 'mockKeyInfo',
+    // Used for testing the testing framework.
+    'GET:/mocked/request' => 'mockTestRequest',
+  ];
+
+  /**
+   * Mock requests for email validation.
+   */
+  protected function mockValidateEmail(ParameterBag $body, ParameterBag $header) : ResponseInterface {
+    // Nothing to do here. In testing we don't really send email.
+    // Verification code "42" works.
+    return $this->success([]);
+  }
+
+  /**
+   * Mock requests for code-based sign-in.
+   */
+  protected function mockSignIn(ParameterBag $body, ParameterBag $header): ResponseInterface {
+    if (!($body->get('username') === 'john@doe.com' && $body->get('code') === '42')) {
+      $this->error(401, 'Invalid code.');
+    }
+    return $this->success([
+      'access_token' => '1234',
+      'token_type' => 'bearer',
+    ]);
+  }
+
+  /**
+   * Mock the user info request.
+   */
+  protected function mockMe(ParameterBag $body, ParameterBag $header) : ResponseInterface {
+    if ($err = $this->authorizeAccess($body, $header)) {
+      return $err;
+    }
+    return $this->success([
+      "email" => "john@doe.com",
+      "id" => 0,
+      "is_active" => TRUE,
+      "is_admin" => TRUE,
+      "team_id" => 1,
+      "team_name" => "amazee.io",
+      "role" => "gm",
+    ]);
+  }
+
+  /**
+   * Verify key access.
+   */
+  protected function authorizeAccess(ParameterBag $body, ParameterBag $header): ?ResponseInterface {
+    if (!$header->has('Authorization')) {
+      $this->error(401, 'Missing authorization header.');
+    }
+    if (!in_array($header->get('Authorization'), ['Bearer 1234', 'Bearer 4321'])) {
+      $this->error(403, 'Invalid authorization header.');
+    }
+    return NULL;
+  }
+
+  /**
+   * Mocked region list for testing.
+   */
+  const REGIONS = [
+    [
+      "id" => 0,
+      "name" => "US 1",
+      "postgres_host" => "https://amazeeio.vdb/us1",
+      "litellm_api_url" => "https://amazeeio.llm/us1",
+      "is_active" => TRUE,
+      "created_at" => "2025-05-12T18:56:01.272Z",
+    ],
+    [
+      "id" => 1,
+      "name" => "Inactive",
+      "postgres_host" => "https://amazeeio.vdb/inactive",
+      "litellm_api_url" => "https://amazeeio.llm/inactive",
+      "is_active" => FALSE,
+      "created_at" => "2025-05-12T18:56:01.272Z",
+    ],
+    [
+      "id" => 2,
+      "name" => "CH 1",
+      "postgres_host" => "https://amazeeio.vdb/ch1",
+      "litellm_api_url" => "https://amazeeio.llm/ch1",
+      "is_active" => TRUE,
+      "created_at" => "2025-05-12T18:56:01.272Z",
+    ],
+  ];
+
+  /**
+   * Mock result of the regions list.
+   */
+  protected function mockRegions(ParameterBag $body, ParameterBag $header): ResponseInterface {
+    if ($err = $this->authorizeAccess($body, $header)) {
+      return $err;
+    }
+    return $this->success(static::REGIONS);
+  }
+
+  /**
+   * Mock private key generation.
+   */
+  protected function mockPostPrivateKey(ParameterBag $body, ParameterBag $header): ResponseInterface {
+    if ($err = $this->authorizeAccess($body, $header)) {
+      return $err;
+    }
+    if (!$body->has('region_id')) {
+      $this->error(400, 'Missing region_id.');
+    }
+    $region_id = $body->get('region_id');
+    if ($region_id === "0") {
+      // Simulate a broken region for error handling tests.
+      $this->error(500, 'Region US 1 is not great again.');
+    }
+    if (!array_key_exists($region_id, static::REGIONS)) {
+      $this->error(400, "Invalid region_id $region_id.");
+    }
+    $region = static::REGIONS[$region_id];
+    return $this->success([
+      'litellm_token' => '4321',
+      'litellm_api_url' => $region['litellm_api_url'],
+    ]);
+  }
+
+  /**
+   * Mock private key list.
+   */
+  protected function mockGetPrivateKeys(ParameterBag $body, ParameterBag $header): ResponseInterface {
+    if ($err = $this->authorizeAccess($body, $header)) {
+      return $err;
+    }
+    return $this->success([
+      [
+        "id" => 0,
+        "name" => "some other key",
+        "region" => static::REGIONS[0]['name'],
+        "database_host" => static::REGIONS[0]['postgres_host'],
+        "database_name" => "db_name_us1",
+        "database_username" => "db_user_us1",
+        "database_password" => "db_pass_us1",
+        "litellm_api_url" => static::REGIONS[0]['litellm_api_url'],
+        "litellm_token" => "5678",
+        "created_at" => "2025-05-13T05:42:48.124Z",
+        "owner_id" => 0,
+        "team_id" => 0,
+      ],
+      [
+        "id" => 1,
+        "name" => AmazeeioAiConfigForm::generatePrivateKeyName(),
+        "region" => static::REGIONS[2]['name'],
+        "database_host" => static::REGIONS[2]['postgres_host'],
+        "database_name" => "db_name_ch1",
+        "database_username" => "db_user_ch1",
+        "database_password" => "db_pass_ch1",
+        "litellm_api_url" => static::REGIONS[2]['litellm_api_url'],
+        "litellm_token" => "4321",
+        "created_at" => "2025-05-13T05:42:48.124Z",
+        "owner_id" => 0,
+        "team_id" => 0,
+      ],
+    ]);
+  }
+
+  /**
+   * Mock key info request.
+   */
+  protected function mockKeyInfo(ParameterBag $body, ParameterBag $header): ResponseInterface {
+    if ($err = $this->authorizeAccess($body, $header)) {
+      return $err;
+    }
+    $key = substr($header->get('Authorization'), strlen('Bearer '));
+    return $this->success([
+      'key' => $key,
+      'info' => [
+        'key_alias' => AmazeeioAiConfigForm::generatePrivateKeyName(),
+        'key_name' => $key,
+        'spend' => 200,
+        'max_budget' => 500,
+        'blocked' => FALSE,
+      ],
+    ]);
+  }
+
+  /**
+   * Mock request for testing the testing framework.
+   */
+  protected function mockTestRequest(ParameterBag $body, ParameterBag $header): ResponseInterface {
+    if (!$header->has('Authentication')) {
+      $this->error(401, 'Missing authentication header');
+    }
+    if (!$body->has('message')) {
+      $this->error(400, 'Missing message argument');
+    }
+    return $this->success([
+      'uppercase' => strtoupper($body->get('message')),
+    ]);
+  }
+
+  /**
+   * Produce a successful response.
+   *
+   * @param array $body
+   *   The JSON response body as an associative array.
+   */
+  protected function success(array $body) : Response {
+    return new Response(200, [], Utils::jsonEncode($body));
+  }
+
+  /**
+   * Throw a client exception.
+   */
+  protected function error(int $status, string $message): void {
+    throw new ClientException(
+      message: $message,
+      request: new Request('GET', ''),
+      response: new Response($status, [], Utils::jsonEncode(['detail' => $message])));
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function request($method, $uri = '', array $options = []): ResponseInterface {
+    $host = parse_url($uri, PHP_URL_HOST);
+    if (strpos($host, 'amazee') === FALSE) {
+      return $this->innerService->get($uri, $options);
+    }
+
+    $path = parse_url($uri, PHP_URL_PATH);
+    $mockKey = "$method:$path";
+    if (array_key_exists($mockKey, $this->requests)) {
+      $method = $this->requests[$mockKey];
+      $body = new ParameterBag(Utils::jsonDecode($options['body'] ?? '{}', TRUE));
+      $headers = new ParameterBag($options['headers'] ?? []);
+      return $this->$method($body, $headers);
+    }
+    throw new ClientException(
+      message: "MockHttpClient: Unhandled request $method:$uri",
+      request: new Request($method, $uri),
+      response: new Response(400),
+    );
+  }
+
+}
diff --git a/src/AmazeeIoApi/AmazeeClient.php b/src/AmazeeIoApi/AmazeeClient.php
index 1969f9a1327766a3ad468b59bef866f5932ab6a0..4e00001b6a88ced5a687d83b5b19ab06bf9e400f 100644
--- a/src/AmazeeIoApi/AmazeeClient.php
+++ b/src/AmazeeIoApi/AmazeeClient.php
@@ -5,6 +5,7 @@ namespace Drupal\ai_provider_amazeeio\AmazeeIoApi;
 use Drupal\Core\Logger\LoggerChannelFactoryInterface;
 use GuzzleHttp\Client;
 use GuzzleHttp\Exception\ClientException;
+use GuzzleHttp\Utils;
 use Psr\Http\Message\ResponseInterface;

 /**
@@ -92,6 +93,36 @@ class AmazeeClient implements ClientInterface {
     return TRUE;
   }

+  /**
+   * Request a validation code for a given email address.
+   */
+  public function requestCode(string $email): void {
+    try {
+      $this->makeRequest('POST', '/auth/validate-email', ['email' => $email]);
+    }
+    catch (ClientException | \Exception $e) {
+      $this->loggerFactory->get('ai_provider_amazeeio')->error('Failed to validate email: @error', ['@error' => $e->getMessage()]);
+    }
+  }
+
+  /**
+   * Validate an email validation code.
+   *
+   * @return ?string
+   *   The access token for this account or null if the code was invalid.
+   */
+  public function validateCode(string $email, string $code): ?string {
+    try {
+      $result = $this->makeRequest('POST', '/auth/sign-in', ['username' => $email, 'code' => $code]);
+      $data = Utils::jsonDecode($result->getBody()->getContents(), TRUE);
+      return $data['access_token'];
+    }
+    catch (ClientException | \Exception $e) {
+      $this->loggerFactory->get('ai_provider_amazeeio')->error('Failed to validate email: @error', ['@error' => $e->getMessage()]);
+    }
+    return NULL;
+  }
+
   /**
    * {@inheritdoc}
    */
@@ -139,7 +170,9 @@ class AmazeeClient implements ClientInterface {
     $region_response = json_decode($response->getBody()->getContents());
     if ($region_response) {
       foreach ($region_response as $region) {
-        $regions[$region->id] = $region->name;
+        if ($region->is_active) {
+          $regions[$region->id] = $region->name;
+        }
       }
     }
     return $regions;
diff --git a/src/AmazeeIoApi/ClientInterface.php b/src/AmazeeIoApi/ClientInterface.php
index 53a82ff5aedc91d5c44fc4ab0a2e7f0485f0b3a7..92b9cedf58d57dafc30e50391e270472e8cd8d97 100644
--- a/src/AmazeeIoApi/ClientInterface.php
+++ b/src/AmazeeIoApi/ClientInterface.php
@@ -44,6 +44,19 @@ interface ClientInterface {
    */
   public function logout(): bool;

+  /**
+   * Request a validation code for a given email address.
+   */
+  public function requestCode(string $email): void;
+
+  /**
+   * Validate an email validation code.
+   *
+   * @return ?string
+   *   The access token for this account or null if the code was invalid.
+   */
+  public function validateCode(string $email, string $code): ?string;
+
   /**
    * Attempt to register and log in to the Amazee API.
    *
diff --git a/src/Form/AmazeeioAiConfigForm.php b/src/Form/AmazeeioAiConfigForm.php
index 57aee45dd616622c4d381ae6fde1ee8cccc5fedf..c81f6f426258af47ed7c247ab0ebaa021e002935 100644
--- a/src/Form/AmazeeioAiConfigForm.php
+++ b/src/Form/AmazeeioAiConfigForm.php
@@ -2,15 +2,13 @@

 namespace Drupal\ai_provider_amazeeio\Form;

+use Drupal\Core\Extension\ModuleHandlerInterface;
 use Drupal\ai\AiProviderPluginManager;
 use Drupal\ai\AiVdbProviderPluginManager;
 use Drupal\ai_provider_amazeeio\AmazeeIoApi\ClientInterface;
 use Drupal\ai_provider_litellm\Form\LiteLlmAiConfigForm;
 use Drupal\ai_provider_litellm\LiteLLM\LiteLlmAiClient;
 use Drupal\ai_provider_openai\OpenAiHelper;
-use Drupal\Component\Render\FormattableMarkup;
-use Drupal\Component\Utility\Html;
-use Drupal\Core\Config\Config;
 use Drupal\Core\Entity\EntityTypeManagerInterface;
 use Drupal\Core\Form\FormStateInterface;
 use Drupal\Core\TempStore\PrivateTempStore;
@@ -45,6 +43,31 @@ class AmazeeioAiConfigForm extends LiteLlmAiConfigForm {
    */
   const POSTGRES_PORT_DEFAULT = '5432';

+  /**
+   * Not connected to amazee.ai.
+   */
+  const STATE_DISCONNECTED = 'disconnected';
+
+  /**
+   * Email address has been entered, waiting for  verification code.
+   */
+  const STATE_VERIFICATION = 'validation';
+
+  /**
+   * Email verification successful, region selection.
+   */
+  const STATE_VERIFIED = 'validated';
+
+  /**
+   * Region has been selected, keys are generated, everything is set up.
+   */
+  const STATE_CONNECTED = 'connected';
+
+  /**
+   * Show a confirmation step before disconnecting.
+   */
+  const STATE_CONFIRM_DISCONNECT = 'confirm_disconnect';
+
   /**
    * Constructs a new LiteLlmAiConfigForm object.
    */
@@ -57,8 +80,12 @@ class AmazeeioAiConfigForm extends LiteLlmAiConfigForm {
     protected PrivateTempStoreFactory $tempStoreFactory,
     protected EntityTypeManagerInterface $entityTypeManager,
     protected AiVdbProviderPluginManager $vdbProviderPluginManager,
+    protected ModuleHandlerInterface $moduleHandler,
   ) {
     parent::__construct($aiProviderManager, $keyRepository, $openAiHelper, $client);
+    $config = $this->config(static::CONFIG_NAME);
+    $this->amazeeClient->setHost($config->get('amazee_host') ?? '');
+    $this->amazeeClient->setToken($this->getTempStore()->get('access_token') ?? '');
   }

   /**
@@ -74,6 +101,7 @@ class AmazeeioAiConfigForm extends LiteLlmAiConfigForm {
       $container->get('tempstore.private'),
       $container->get('entity_type.manager'),
       $container->get('ai.vdb_provider'),
+      $container->get('module_handler')
     );
   }

@@ -85,557 +113,349 @@ class AmazeeioAiConfigForm extends LiteLlmAiConfigForm {
   }

   /**
-   * {@inheritdoc}
-   */
-  public function buildForm(array $form, FormStateInterface $form_state): array {
-    $config = $this->config(static::CONFIG_NAME);
-
-    $form['#id'] = Html::getUniqueId('amazee-config-form');
-    // Add the custom Amazee section.
-    $form['amazee'] = [
-      '#type' => 'details',
-      '#title' => $this->t('Amazee.io login'),
-      '#attributes' => [
-        // ID used for AJAX callback wrapper.
-        'id' => 'amazee-details',
-      ],
-      '#open' => TRUE,
-    ];
-
-    if ($auth_token = $this->getTempStore()->get('access_token')) {
-      $this->amazeeClient->setToken($auth_token);
-    }
-    $this->amazeeClient->setHost($config->get('amazee_host') ?? '');
-
-    $current_key = $this->keyRepository->getKey('amazeeio_ai')?->getKeyValue();
-    if (!$auth_token) {
-      $this->loginRegisterForm($form['amazee'], $form_state, $form['#id'], empty($current_key));
-    }
-    else {
-      $authorized = $this->amazeeClient->authorized();
-      if (!$authorized) {
-        $this->loginRegisterForm($form['amazee'], $form_state, $form['#id'], empty($current_key));
-        // Clear the token as it no longer works.
-        $this->getTempStore()->delete('access_token');
-        $this->amazeeClient->setToken('');
-      }
-      else {
-        // We don't need to see the Amazee.io login/register if authorized.
-        $form['amazee']['#access'] = FALSE;
-      }
-    }
-
-    $this->apiKeysForm($form, $form_state, $authorized ?? FALSE);
-
-    // Show key usage details.
-    $this->keyUsageForm(
-      $form,
-      $form_state,
-      $config,
-    );
-
-    return $form;
-  }
-
-  /**
-   * Add Login/Register subform to the form.
+   * Determine the current form state.
    *
-   * @param array $sub_form
-   *   An array of form elements to add the subform to.
-   * @param \Drupal\Core\Form\FormStateInterface $form_state
-   *   The current form state.
-   * @param string $form_id
-   *   The Form ID.
-   * @param bool $open
-   *   Whether the details elements should be open.
+   * Based on the current `$form_state` as well as the authorization
+   * status.
    */
-  protected function loginRegisterForm(array &$sub_form, FormStateInterface $form_state, string $form_id, bool $open = FALSE): void {
-    // Integrate with the Amazee AI Keys as a Service API.
-    $sub_form['login'] = [
-      '#type' => 'details',
-      '#title' => $this->t('Log in'),
-      '#open' => $open,
-      'login_username' => [
-        '#type' => 'textfield',
-        '#title' => $this->t('Username'),
-      ],
-      'login_password' => [
-        '#type' => 'password',
-        '#title' => $this->t('Password'),
-      ],
-      'login_button' => [
-        '#type' => 'submit',
-        '#value' => $this->t('Log in'),
-        '#validate' => ['::loginValidate'],
-        '#submit' => ['::loginSubmit'],
-        '#ajax' => [
-          'callback' => '::wholeFormAjax',
-          'wrapper' => $form_id,
-        ],
-        '#attributes' => ['class' => ['button', 'button--primary']],
-      ],
-    ];
-
-    $sub_form['register'] = [
-      '#type' => 'details',
-      '#title' => $this->t('Register'),
-      '#open' => $open,
-      'register_username' => [
-        '#type' => 'textfield',
-        '#title' => $this->t('Username'),
-      ],
-      'register_password' => [
-        '#type' => 'password',
-        '#title' => $this->t('Password'),
-      ],
-      'register_button' => [
-        '#type' => 'button',
-        '#value' => $this->t('Register'),
-        '#validate' => ['::registerValidate'],
-        '#submit' => ['::registerSubmit'],
-        '#ajax' => [
-          'callback' => '::wholeFormAjax',
-          'wrapper' => $form_id,
-        ],
-        '#attributes' => ['class' => ['button', 'button--primary']],
-      ],
-    ];
+  public function currentState(FormStateInterface $form_state) : string {
+    return $form_state->get('state')
+      ? $form_state->get('state')
+      : ($this->amazeeClient->authorized()
+        ? static::STATE_CONNECTED
+        : static::STATE_DISCONNECTED
+      );
   }

   /**
-   * Add Log out subform to the form.
-   *
-   * @param array $sub_form
-   *   An array of form elements to add the subform to.
-   * @param \Drupal\Core\Form\FormStateInterface $form_state
-   *   The current form state.
+   * Determine if the module is in "test mode".
    */
-  protected function logoutForm(array &$sub_form, FormStateInterface $form_state): void {
-    $sub_form['logout_button'] = [
-      '#type' => 'submit',
-      '#value' => $this->t('Log out'),
-      '#validate' => [],
-      '#submit' => ['::logoutSubmit'],
-      '#attributes' => ['class' => ['button', 'button--danger']],
-    ];
+  protected function testMode() : bool {
+    return $this->moduleHandler->moduleExists('ai_provider_amazeeio_test');
   }

   /**
-   * Add API key usage subform to the form.
-   *
-   * @param array<string, mixed> $form
-   *   An array of form elements to add the subform to.
-   * @param \Drupal\Core\Form\FormStateInterface $form_state
-   *   The current form state.
+   * {@inheritdoc}
    */
-  protected function selectKeyForm(array &$form, FormStateInterface $form_state): void {
-    $key_options = $this->getApiKeys();
-    $form['api_key'] = [
-      '#type' => 'select',
-      '#title' => $this->t('API key'),
-      '#description' => $this->t('Select an existing API key to use.'),
-      '#options' => $key_options,
-      '#default_value' => $this->keyRepository->getKey(static::API_KEY_NAME)?->getKeyValue(),
-      '#required' => TRUE,
-      '#access' => count($key_options) > 1,
-    ];
+  public function buildForm(array $form, FormStateInterface $form_state): array {
+    $config = $this->config(static::CONFIG_NAME);
+    $this->amazeeClient->setToken($this->getTempStore()->get('access_token') ?? '');

-    $form['api_key_none'] = [
-      '#markup' => $this->t('<p>You currently do not have any API keys.</p>'),
-      '#access' => count($key_options) === 1,
-    ];
+    if ($auth_token = $this->getTempStore()->get('access_token')) {
+      $this->amazeeClient->setToken($auth_token);
+    }

-    $form['show_generate'] = [
-      '#type' => 'submit',
-      '#value' => $this->t('Generate new key'),
-      '#validate' => ['::showGenerateValidate'],
-      '#submit' => ['::showGenerateSubmit'],
-      '#limit_validation_errors' => [],
-      '#ajax' => [
-        'callback' => '::wholeFormAjax',
-        'wrapper' => $form['#id'],
+    $buttonAjax = [
+      'callback' => '::ajaxUpdate',
+      'event' => 'click',
+      'wrapper' => 'amazee-ai-config-form',
+      'progress' => [
+        'type' => 'throbber',
       ],
     ];

-    $form['submit'] = [
-      '#type' => 'submit',
-      '#value' => $this->t('Save configuration'),
-      '#attributes' => ['class' => ['button', 'button--primary']],
-      '#access' => count($key_options) !== 1,
+    $state = $this->currentState($form_state);
+    $form['image'] = [
+      "#markup" => '<p><img src="http://assets.amazee.ai/logo.png" alt="amazee.ai" width="400"/>',
     ];
-  }
-
-  /**
-   * Add key and usage info for the current key to a form.
-   *
-   * @param array $form
-   *   The form to add to.
-   * @param \Drupal\Core\Form\FormStateInterface $form_state
-   *   The current form state.
-   * @param \Drupal\Core\Config\Config $config
-   *   The Amazee.io AI config.
-   */
-  protected function keyUsageForm(array &$form, FormStateInterface $form_state, Config $config): void {
-    $host = $config->get('host');
-    if (empty($host) || !$this->keyRepository->getKey(static::API_KEY_NAME)->getKeyValue()) {
-      // Show nothing if we don't have an API key or host yet.
-      return;
-    }
-
-    $client = new LiteLlmAiClient($this->client, $this->keyRepository, $host, 'amazeeio_ai');
-    $keys = $client->keyInfo();
-    $key_info = reset($keys);
-
-    $form['usage'] = [
-      '#theme' => 'table',
-      '#rows' => [],
-      '#weight' => 20,
+    $ajax = [
+      '#prefix' => '<div id="amazee-ai-config-form">',
+      '#suffix' => '</div>',
     ];
-
-    if ($key_info->info->key_alias) {
-      $form['usage']['#rows'][] = [
-        $this->t('Name'),
-        $key_info->info->key_alias,
+    if ($state === static::STATE_DISCONNECTED) {
+      $ajax['markup'] = [
+        '#markup' => '<p><em>' . $this->t('Enter your email address to receive a verification code to connect to <strong>amazee.ai</strong>.') . '</em></p>',
       ];
-    }
-
-    $form['usage']['#rows'][] = [
-      $this->t('Key'),
-      $key_info->info->key_name,
-    ];
-
-    $form['usage']['#rows'][] = [
-      $this->t('Spend ($)'),
-      number_format($key_info->info->spend, 5),
-    ];
-
-    $form['usage']['#rows'][] = [
-      $this->t('Max budget ($)'),
-      $key_info->info->max_budget === NULL ? $this->t('N/A') : number_format($key_info->info->max_budget, 2),
-    ];
-
-    $form['usage']['#rows'][] = [
-      $this->t('Blocked'),
-      $key_info->info->blocked ? $this->t('Yes') : $this->t('No'),
-    ];
-
-    if ($database = $config->get('postgres_default_database')) {
-      $form['usage']['#rows'][] = [
-        $this->t('VectorDB Database'),
-        $database,
+      $ajax['email'] = [
+        // When in 'test mode' we use a simple text field, so the BrowserTest
+        // is actually able to enter an invalid email address.
+        '#type' => $this->testMode() ? 'textfield' : 'email',
+        '#title' => $this->t('Email'),
+        '#description' => $this->t('By entering your email address, you agree to amazee.io\'s <a href="https://www.amazee.ai/terms-of-service">Terms of Service.</a>'),
+      ];
+      $ajax['submit_email'] = [
+        '#type' => 'submit',
+        '#value' => $this->t('Sign in'),
+        '#ajax' => $buttonAjax,
+        '#attributes' => ['class' => ['button', 'button--primary']],
       ];
     }

-    foreach ($form['usage']['#rows'] as &$row) {
-      $row[0] = [
-        'data' => ['#markup' => $row[0]],
-        'header' => TRUE,
+    if ($state === static::STATE_VERIFICATION) {
+      $ajax['markup'] = [
+        '#markup' => '<p><em>' . $this->t('Enter the verification code that you should have received via email.') . '</em></p>',
+      ];
+      $ajax['code'] = [
+        '#type' => 'textfield',
+        '#title' => $this->t('Code'),
+      ];
+      $ajax['submit_code'] = [
+        '#type' => 'submit',
+        '#value' => $this->t('Validate'),
+        '#ajax' => $buttonAjax,
+        '#attributes' => ['class' => ['button', 'button--primary']],
       ];
     }
-  }

-  /**
-   * Add API keys section of the form (selecting existing & generating new).
-   *
-   * @param array $form
-   *   The current form.
-   * @param \Drupal\Core\Form\FormStateInterface $form_state
-   *   The current form state.
-   * @param bool $authorized
-   *   Whether the user is currently authorized with the Amazee API.
-   */
-  protected function apiKeysForm(array &$form, FormStateInterface $form_state, bool $authorized): void {
-    $generate_new = $form_state->get('amazee_show_generate') ?? FALSE;
-    if ($generate_new) {
-      $this->generateApiKeyForm($form, $form_state);
-    }
-    elseif ($authorized) {
-      $this->selectKeyForm($form, $form_state);
-    }
-    if ($authorized) {
-      $this->logoutForm($form, $form_state);
+    if ($state === static::STATE_VERIFIED) {
+      try {
+        $regions = $this->amazeeClient->getRegions();
+      }
+      catch (ClientException $e) {
+        $this->messenger->addError($this->t('An error occurred while retrieving the available regions. Please consult the Drupal error log.'));
+      }
+      $ajax['markup'] = [
+        '#markup' => '<p><em>' . $this->t('Select a region to host AI features in.') . '</em></p>',
+      ];
+      $ajax['region'] = [
+        '#type' => 'select',
+        '#title' => $this->t('Region'),
+        '#options' => $regions ?? [],
+        '#title_display' => 'before',
+        '#access' => !empty($regions),
+      ];
+      $ajax['submit_region'] = [
+        '#type' => 'submit',
+        '#value' => $this->t('Connect'),
+        '#access' => !empty($regions),
+        '#attributes' => ['class' => ['button', 'button--primary']],
+      ];
     }
-  }

-  /**
-   * Add generate API key subform to the form.
-   *
-   * @param array $form
-   *   An array of form elements to add the subform to.
-   * @param \Drupal\Core\Form\FormStateInterface $form_state
-   *   The current form state.
-   */
-  protected function generateApiKeyForm(array &$form, FormStateInterface $form_state): void {
-    $form['key_name'] = [
-      '#type' => 'textfield',
-      '#title' => $this->t('Name'),
-      '#description' => $this->t('A name for the key to help identify it.'),
-    ];
+    if ($state === static::STATE_CONNECTED) {
+      $ajax['submit_disconnect'] = [
+        '#type' => 'submit',
+        '#value' => $this->t('Disconnect'),
+        '#attributes' => ['class' => ['button', 'button--danger']],
+      ];

-    try {
-      $regions = $this->amazeeClient->getRegions();
+      $host = $config->get('host');
+      if (!(empty($host) || !$this->keyRepository->getKey(static::API_KEY_NAME)->getKeyValue())) {
+        $client = new LiteLlmAiClient($this->client, $this->keyRepository, $host, 'amazeeio_ai');
+        $keys = $client->keyInfo();
+        $key_info = reset($keys);
+
+        $ajax['usage'] = [
+          '#theme' => 'table',
+          '#rows' => [],
+          '#weight' => 20,
+        ];
+
+        if ($key_info->info->key_alias) {
+          $ajax['usage']['#rows'][] = [
+            $this->t('Name'),
+            $key_info->info->key_alias,
+          ];
+        }
+
+        if ($database = $config->get('postgres_default_database')) {
+          $ajax['usage']['#rows'][] = [
+            $this->t('VectorDB Database'),
+            $database,
+          ];
+        }
+
+        foreach ($ajax['usage']['#rows'] as &$row) {
+          $row[0] = [
+            'data' => ['#markup' => $row[0]],
+            'header' => TRUE,
+          ];
+        }
+      }
     }
-    catch (ClientException $e) {
-      // Do nothing as form will handle empty regions array.
+    if ($state === static::STATE_CONFIRM_DISCONNECT) {
+      $ajax['markup'] = [
+        '#markup' => '<p><em>' . $this->t('Do you really want to disconnect from <strong>amazee.ai</string>?') . '</em></p>',
+      ];
+      $ajax['submit_confirm_disconnect'] = [
+        '#type' => 'submit',
+        '#value' => $this->t('Disconnect'),
+        '#attributes' => ['class' => ['button', 'button--danger']],
+      ];
+      $ajax['cancel'] = [
+        '#type' => 'submit',
+        '#value' => $this->t('No! Go back!'),
+        '#ajax' => $buttonAjax,
+        '#attributes' => ['class' => ['button', 'button--secondary']],
+      ];
     }
-    $form['region'] = [
-      '#type' => 'select',
-      '#title' => $this->t('Region'),
-      '#options' => $regions ?? [],
-      '#access' => !empty($regions),
-    ];
+    $form['ajax'] = $ajax;

-    $form['generate_key'] = [
-      '#type' => 'submit',
-      '#value' => $this->t('Generate API key'),
-      '#validate' => ['::generateKeyValidate'],
-      '#submit' => ['::generateKeySubmit'],
-      '#ajax' => [
-        'callback' => '::wholeFormAjax',
-        'wrapper' => $form['#id'],
-      ],
-    ];
+    return $form;
   }

   /**
-   * AJAX handler for Amazee section of the form.
-   *
-   * @param array<string, mixed> $form
-   *   The form being submitted.
-   * @param \Drupal\Core\Form\FormStateInterface $form_state
-   *   The form state being submitted.
-   *
-   * @return array<string, mixed>
-   *   The Amazee-specific section of the form.
+   * Ajax callback to dynamically update the form.
    */
-  public function amazeeAjax(array $form, FormStateInterface $form_state): array {
-    return $form['amazee'];
+  public static function ajaxUpdate(array &$form, FormStateInterface $form_state) {
+    return $form['ajax'];
   }

   /**
-   * Validation handler for Login action.
-   *
-   * @param array<string, mixed> $form
-   *   The form being submitted.
-   * @param \Drupal\Core\Form\FormStateInterface $form_state
-   *   The form state being submitted.
+   * Signup form validation.
    */
-  public function loginValidate(array &$form, FormStateInterface $form_state): void {
-    $config = $this->config(static::CONFIG_NAME);
-    $this->amazeeClient->setHost($config->get('amazee_host') ?? '');
-    $access_token = $this->amazeeClient->login($form_state->getValue('login_username'), $form_state->getValue('login_password'));
-    if (empty($access_token)) {
-      $form_state->setError($form['amazee']['login'], $this->t('Could not log in to Amazee.io. Please check your details.'));
-      return;
+  public function validateForm(array &$form, FormStateInterface $form_state): void {
+    $state = $this->currentState($form_state);
+
+    if ($state === static::STATE_DISCONNECTED) {
+      $email = $form_state->getValue('email');
+      if (filter_var($email, FILTER_VALIDATE_EMAIL)) {
+        $form_state->set('email', $email);
+        $this->amazeeClient->requestCode($email);
+        $form_state->set('state', static::STATE_VERIFICATION);
+      }
+      else {
+        $form_state->setErrorByName('email', $this->t('Invalid email address.'));
+      }
     }
-
-    $this->getTempStore()->set('access_token', $access_token);
-  }
-
-  /**
-   * Submit handler for Log in action.
-   *
-   * @param array<string, mixed> $form
-   *   The form being submitted.
-   * @param \Drupal\Core\Form\FormStateInterface $form_state
-   *   The form state being submitted.
-   */
-  public function loginSubmit(array &$form, FormStateInterface $form_state): void {
-    $form_state->setRebuild();
-  }
-
-  /**
-   * Submit handler for Log out action.
-   *
-   * @param array<string, mixed> $form
-   *   The form being submitted.
-   * @param \Drupal\Core\Form\FormStateInterface $form_state
-   *   The form state being submitted.
-   */
-  public function logoutSubmit(array &$form, FormStateInterface $form_state): void {
-    $this->amazeeClient->logout();
-    $this->getTempStore()->delete('access_token');
-    $form_state->setRedirect($this->getRouteMatch()->getRouteName(), $this->getRouteMatch()->getParameters()->all());
-  }
-
-  /**
-   * Validation handler for Register action.
-   *
-   * @param array<string, mixed> $form
-   *   The form being submitted.
-   * @param \Drupal\Core\Form\FormStateInterface $form_state
-   *   The form state being submitted.
-   */
-  public function registerValidate(array &$form, FormStateInterface $form_state): void {
-    $access_token = $this->amazeeClient->register($form_state->getValue('register_username'), $form_state->getValue('register_password'));
-    if (empty($access_token)) {
-      $form_state->setError($form['amazee']['register'], $this->t('Could not register with Amazee.io.'));
-      return;
+    if ($state === static::STATE_VERIFICATION) {
+      $email = $form_state->get('email');
+      $code = $form_state->getValue('code');
+      $token = $this->amazeeClient->validateCode($email, $code);
+      if ($token) {
+        $this->getTempStore()->set('access_token', $token);
+        $form_state->set('state', static::STATE_VERIFIED);
+      }
+      else {
+        $form_state->setErrorByName('code', $this->t('The provided code is incorrect or has expired.'));
+      }
     }
-
-    $this->getTempStore()->set('access_token', $access_token);
-  }
-
-  /**
-   * Submit handler for Register action.
-   *
-   * @param array<string, mixed> $form
-   *   The form being submitted.
-   * @param \Drupal\Core\Form\FormStateInterface $form_state
-   *   The form state being submitted.
-   */
-  public function registerSubmit(array &$form, FormStateInterface $form_state): void {
-    $form_state->setRebuild();
-  }
-
-  /**
-   * Validation handler for show generate key form.
-   *
-   * @param array<string, mixed> $form
-   *   The current form.
-   * @param \Drupal\Core\Form\FormStateInterface $form_state
-   *   The current form state.
-   */
-  public function showGenerateValidate(array &$form, FormStateInterface $form_state): void {
-    // Intentionally left blank.
-  }
-
-  /**
-   * Submit handler for showing the generate key form.
-   *
-   * @param array<string, mixed> $form
-   *   The current form.
-   * @param \Drupal\Core\Form\FormStateInterface $form_state
-   *   The current form state.
-   */
-  public function showGenerateSubmit(array &$form, FormStateInterface $form_state): void {
-    $form_state->setRebuild();
-    $form_state->set('amazee_show_generate', TRUE);
-  }
-
-  /**
-   * AJAX callback to replace the whole form.
-   *
-   * @param array<string, mixed> $form
-   *   The current form.
-   * @param \Drupal\Core\Form\FormStateInterface $form_state
-   *   The current form state.
-   *
-   * @return array<string, mixed>
-   *   The whole form.
-   */
-  public function wholeFormAjax(array $form, FormStateInterface $form_state): array {
-    return $form;
-  }
-
-  /**
-   * Validation handler for Generate Key action.
-   *
-   * @param array<string, mixed> $form
-   *   The form being submitted.
-   * @param \Drupal\Core\Form\FormStateInterface $form_state
-   *   The form state being submitted.
-   */
-  public function generateKeyValidate(array &$form, FormStateInterface $form_state): void {
-    $region = $form_state->getValue('region');
-    if (empty($region)) {
-      $form_state->setError($form['amazee']['region'], $this->t('Unable to fetch list of regions from Amazee.io.'));
-      return;
+    if ($state === static::STATE_VERIFIED) {
+      $region = $form_state->getValue('region');
+      $private_key = $this->amazeeClient->createPrivateAiKey(
+        $region,
+        static::generatePrivateKeyName()
+      );
+      if (!$private_key) {
+        $form_state->setErrorByName('region', $this->t('An error occurred while generating the private key. Please consult the Drupal error log.'));
+      }
+      else {
+        // Return now to not rebuild the form but submit it.
+        return;
+      }
     }
-
-    $key_name = $form_state->getValue('key_name');
-    $access_token = $this->getTempStore()->get('access_token');
-    if (empty($access_token)) {
-      $form_state->setError($form['amazee'], $this->t('You have been logged out. Please log in and try again.'));
-      return;
+    if ($state === static::STATE_CONNECTED) {
+      $element = $form_state->getTriggeringElement();
+      if ($element['#id'] === 'edit-submit-disconnect') {
+        $form_state->set('state', static::STATE_CONFIRM_DISCONNECT);
+      }
     }
-    $config = $this->config(static::CONFIG_NAME);
-
-    $this->amazeeClient->setToken($access_token);
-    $this->amazeeClient->setHost($config->get('amazee_host') ?? '');
-
-    $private_key = $this->amazeeClient->createPrivateAiKey($region, $key_name);
-    // If successful remove the generate form.
-    if ($private_key['litellm_token'] && $private_key['litellm_api_url']) {
-      $form_state->set('amazee_show_generate', FALSE);
+    if ($state === static::STATE_CONFIRM_DISCONNECT) {
+      $element = $form_state->getTriggeringElement();
+      if ($element['#id'] === 'edit-submit-confirm-disconnect') {
+        $form_state->set('state', static::STATE_CONFIRM_DISCONNECT);
+        // Return now to not rebuild the form but submit it.
+        return;
+      }
+      else {
+        $form_state->set('state', static::STATE_CONNECTED);
+      }
     }
     $form_state->setRebuild();
   }

   /**
-   * Submit handler for Generate Key action.
+   * Generate a key name for this installation.
    *
-   * @param array<string, mixed> $form
-   *   The form being submitted.
-   * @param \Drupal\Core\Form\FormStateInterface $form_state
-   *   The form state being submitted.
+   * Assumes that each Drupal installation has a single API key.
    */
-  public function generateKeySubmit(array &$form, FormStateInterface $form_state): void {
-    $form_state->setRebuild();
-  }
-
-  /**
-   * {@inheritdoc}
-   */
-  public function validateForm(array &$form, FormStateInterface $form_state): void {
-    // Override parent with intentionally left blank as the flow for this form
-    // validates API key selection elsewhere.
+  public static function generatePrivateKeyName(): string {
+    return \Drupal::request()->getHost();
   }

   /**
    * {@inheritdoc}
    */
   public function submitForm(array &$form, FormStateInterface $form_state): void {
-    $config = $this->config(static::CONFIG_NAME);
-    $this->amazeeClient->setToken($this->getTempStore()->get('access_token') ?? '');
-    $this->amazeeClient->setHost($config->get('amazee_host') ?? '');
-    $api_key = $this->amazeeClient->getPrivateApiKey($form_state->getValue('api_key'));
-    if ($api_key) {
-      // Set the provider config, using a known key name to ease support
-      // preconfigured environments.
+    if ($form_state->get('state') === static::STATE_VERIFIED) {
+      $config = $this->config(static::CONFIG_NAME);
+      $this->amazeeClient->setToken($this->getTempStore()->get('access_token') ?? '');
+      $this->amazeeClient->setHost($config->get('amazee_host') ?? '');
+      $key_name = static::generatePrivateKeyName();
+      $api_keys = array_filter(
+        $this->amazeeClient->getPrivateApiKeys(),
+        fn($key) => $key->name === $key_name
+      );
+      $api_key = reset($api_keys);
+
+      if ($api_key) {
+        // Set the provider config, using a known key name to ease support
+        // preconfigured environments.
+        $this->config(static::CONFIG_NAME)
+          ->set('host', $api_key->litellm_api_url)
+          ->set('postgres_host', $api_key->database_host)
+          ->set('postgres_port', $api_key->database_port ?? static::POSTGRES_PORT_DEFAULT)
+          ->set('postgres_default_database', $api_key->database_name)
+          ->set('postgres_username', $api_key->database_username)
+          ->set('postgres_password', static::VDB_PASSWORD_NAME)
+          ->set('api_key', static::API_KEY_NAME)
+          ->save();
+
+        // Load or create the Amazee.io key.
+        /** @var \Drupal\Core\Entity\EntityStorageInterface $key_storage */
+        $key_storage = $this->entityTypeManager->getStorage('key');
+        /** @var \Drupal\key\Entity\Key $key*/
+        $key = $key_storage->load(static::API_KEY_NAME) ??
+          $key_storage->create([
+            'id' => static::API_KEY_NAME,
+            'label' => 'Amazee.io AI API Key',
+            'description' => 'Automatically created by the Amazee.io AI provider.',
+          ]);
+        // Update the key config.
+        $key
+          ->set('key_provider', 'config')
+          ->set('key_provider_settings', ['key_value' => $api_key->litellm_token])
+          ->set('key_input', 'text_field')
+          ->save();
+
+        // Load or create the Amazee.io Postgres key.
+        /** @var \Drupal\key\Entity\Key $database_key */
+        $database_key = $key_storage->load(static::VDB_PASSWORD_NAME) ??
+          $key_storage->create([
+            'id' => static::VDB_PASSWORD_NAME,
+            'label' => 'Amazee.io AI Database Key',
+            'description' => 'Automatically created by the Amazee.io AI provider.',
+          ]);
+        // Update the key config.
+        $database_key
+          ->set('key_provider', 'config')
+          ->set('key_provider_settings', ['key_value' => $api_key->database_password])
+          ->set('key_input', 'text_field')
+          ->save();
+
+        // Set the default models where available.
+        /** @var \Drupal\ai_provider_amazeeio\Plugin\AiProvider\AmazeeioAiProvider $provider */
+        $provider = $this->aiProviderManager->createInstance('amazeeio');
+        // Run post-setup when not in unit tests, since it connects to the
+        // real LLM.
+        if (!$this->testMode()) {
+          $provider->postSetup();
+        }
+        $this->messenger()->addStatus($this->t('This website has been connected to <strong>amazee.ai</strong>.'));
+      }
+    }
+    if ($form_state->get('state') === static::STATE_CONFIRM_DISCONNECT) {
+      $this->getTempStore()->delete('access_token');
       $this->config(static::CONFIG_NAME)
-        ->set('host', $api_key->litellm_api_url)
-        ->set('postgres_host', $api_key->database_host)
-        ->set('postgres_port', $api_key->database_port ?? static::POSTGRES_PORT_DEFAULT)
-        ->set('postgres_default_database', $api_key->database_name)
-        ->set('postgres_username', $api_key->database_username)
+        ->set('host', '')
+        ->set('postgres_host', '')
+        ->set('postgres_port', static::POSTGRES_PORT_DEFAULT)
+        ->set('postgres_default_database', '')
+        ->set('postgres_username', '')
         ->set('postgres_password', static::VDB_PASSWORD_NAME)
         ->set('api_key', static::API_KEY_NAME)
         ->save();
-
-      // Load or create the Amazee.io key.
+      /** @var EntityStorageInterface $key_storage */
       $key_storage = $this->entityTypeManager->getStorage('key');
-      $key = $key_storage->load(static::API_KEY_NAME) ??
-        $key_storage->create([
-          'id' => static::API_KEY_NAME,
-          'label' => 'Amazee.io AI API Key',
-          'description' => 'Automatically created by the Amazee.io AI provider.',
-        ]);
-      // Update the key config.
-      $key
-        ->set('key_provider', 'config')
-        ->set('key_provider_settings', ['key_value' => $api_key->litellm_token])
-        ->set('key_input', 'text_field')
-        ->save();
-
-      // Load or create the Amazee.io Postgres key.
-      $database_key = $key_storage->load(static::VDB_PASSWORD_NAME) ??
-        $key_storage->create([
-          'id' => static::VDB_PASSWORD_NAME,
-          'label' => 'Amazee.io AI Database Key',
-          'description' => 'Automatically created by the Amazee.io AI provider.',
-        ]);
-      // Update the key config.
-      $database_key
-        ->set('key_provider', 'config')
-        ->set('key_provider_settings', ['key_value' => $api_key->database_password])
-        ->set('key_input', 'text_field')
-        ->save();
-
-      // Set the default models where available.
-      $this->aiProviderManager
-        ->createInstance('amazeeio')
-        ->postSetup();
+      $apiKey = $key_storage->load(static::API_KEY_NAME);
+      $dbKey = $key_storage->load(static::VDB_PASSWORD_NAME);
+      if ($apiKey) {
+        $apiKey->delete();
+      }
+      if ($dbKey) {
+        $dbKey->delete();
+      }
+      $this->messenger()->addWarning($this->t('This website has been disconnected from <strong>amazee.ai</strong>.'));
     }
   }

@@ -649,28 +469,4 @@ class AmazeeioAiConfigForm extends LiteLlmAiConfigForm {
     return $this->tempStoreFactory->get('amazeeio_ai');
   }

-  /**
-   * Get the available API keys suitable for a select element.
-   *
-   * @return array<string, string>
-   *   The API keys, with the token as the key and the name as the value.
-   *   Includes an empty option.
-   */
-  protected function getApiKeys(): array {
-    $keys = [
-      '' => $this->t('- Select -'),
-    ];
-    foreach ($this->amazeeClient->getPrivateApiKeys() as $private_api_key) {
-      $keys[$private_api_key->litellm_token] = new FormattableMarkup(
-        '@name [@region]',
-        [
-          '@name' => $private_api_key->name ?? substr_replace($private_api_key->litellm_token, '...', 3, -4),
-          '@region' => $private_api_key->region,
-        ],
-      );
-    }
-
-    return $keys;
-  }
-
 }
diff --git a/tests/src/Functional/AmazeeioProviderTest.php b/tests/src/Functional/AmazeeioProviderTest.php
new file mode 100644
index 0000000000000000000000000000000000000000..8d7239425509b6f91941317f374342b7624842d6
--- /dev/null
+++ b/tests/src/Functional/AmazeeioProviderTest.php
@@ -0,0 +1,190 @@
+<?php
+
+namespace Drupal\Tests\ai_provider_amazeeio\Functional;
+
+use Drupal\ai_provider_amazeeio\Form\AmazeeioAiConfigForm;
+
+use Drupal\Tests\BrowserTestBase;
+
+/**
+ * Integration tests for amazee.io AI provider user interfaces.
+ */
+class AmazeeioProviderTest extends BrowserTestBase {
+
+  /**
+   * {@inheritdoc}
+   */
+  protected $defaultTheme = 'stark';
+
+  /**
+   * Disable strict config checking.
+   *
+   * @var bool
+   * @todo Config schema of 'ai_vdb_provider_postgres' is broken.
+   */
+  public $strictConfigSchema = FALSE;
+
+  /**
+   * {@inheritdoc}
+   */
+  protected static $modules = [
+    'ai',
+    'ai_vdb_provider_postgres',
+    'ai_provider_litellm',
+    'ai_provider_amazeeio',
+    'ai_provider_amazeeio_test',
+  ];
+
+  /**
+   * {@inheritdoc}
+   */
+  protected function setUp() : void {
+    parent::setUp();
+    $account = $this->drupalCreateUser([
+      'administer ai',
+      'administer ai providers',
+    ]);
+    $this->drupalLogin($account);
+  }
+
+  /**
+   * Provider is in the list of available providers.
+   */
+  public function testProviderAvailable() {
+    $this->drupalGet('/admin/config/ai/providers');
+    $this->assertSession()->statusCodeEquals(200);
+    $this->assertSession()->linkExists('Amazee.io AI Authentication');
+  }
+
+  /**
+   * Fill the email step.
+   */
+  protected function fillEmail(string $email) {
+    $page = $this->getSession()->getPage();
+    $this->drupalGet('/admin/config/ai/providers/amazeeio');
+    $page->fillField('Email', $email);
+    $page->pressButton('Sign in');
+    $this->assertSession()->statusCodeEquals(200);
+  }
+
+  /**
+   * Fill the code step.
+   */
+  protected function fillCode(string $code) {
+    $page = $this->getSession()->getPage();
+    $page->fillField('Code', $code);
+    $page->pressButton('Validate');
+    $this->assertSession()->statusCodeEquals(200);
+  }
+
+  /**
+   * Select a region.
+   */
+  protected function selectRegion(string $label) {
+    $page = $this->getSession()->getPage();
+    $page->selectFieldOption('Region', $label);
+    $page->pressButton('Connect');
+    $this->assertSession()->statusCodeEquals(200);
+  }
+
+  /**
+   * Email address validation.
+   */
+  public function testInvalidEmail() {
+    $this->fillEmail('john+doe.com');
+    $this->assertSession()->pageTextContains('Invalid email address.');
+  }
+
+  /**
+   * Invalid code error message.
+   */
+  public function testInvalidCode() {
+    $this->fillEmail('john@doe.com');
+    $this->fillCode('666');
+    $this->assertSession()->pageTextContains('The provided code is incorrect or has expired.');
+  }
+
+  /**
+   * Inactive regions are not selectable.
+   */
+  public function testInactiveRegions() {
+    $this->fillEmail('john@doe.com');
+    $this->fillCode('42');
+    $regions = $this->getOptions('Region');
+    // "1" is the index of the "Inactive" region.
+    // @see MockHttpClient::mockRegions
+    $this->assertArrayNotHasKey(1, $regions);
+    $this->assertCount(2, $regions);
+  }
+
+  /**
+   * Error handling when region API breaks.
+   */
+  public function testBrokenRegion() {
+    $this->fillEmail('john@doe.com');
+    $this->fillCode('42');
+    $this->selectRegion('US 1');
+    $this->assertSession()->pageTextContains('An error occurred while generating the private key.');
+  }
+
+  /**
+   * Complete process and resulting keys.
+   */
+  public function testKeyGeneration() {
+    $this->fillEmail('john@doe.com');
+    $this->fillCode('42');
+    $this->selectRegion('CH 1');
+
+    $config = \Drupal::configFactory()->get(AmazeeioAiConfigForm::CONFIG_NAME);
+    $this->assertEquals('https://amazeeio.llm/ch1', $config->get('host'));
+
+    $this->assertEquals("db_name_ch1", $config->get('postgres_default_database'));
+    $this->assertEquals("https://amazeeio.vdb/ch1", $config->get('postgres_host'));
+    $this->assertEquals("amazeeio_ai_database", $config->get('postgres_password'));
+    $this->assertEquals(5432, $config->get('postgres_port'));
+    $this->assertEquals("db_user_ch1", $config->get('postgres_username'));
+
+    $keyStorage = \Drupal::entityTypeManager()->getStorage('key');
+    $aiKey = $keyStorage->load(AmazeeioAiConfigForm::API_KEY_NAME);
+    $dbKey = $keyStorage->load(AmazeeioAiConfigForm::VDB_PASSWORD_NAME);
+    $this->assertEquals('4321', $aiKey->getKeyValue());
+    $this->assertEquals('db_pass_ch1', $dbKey->getKeyValue());
+
+    $this->assertSession()->buttonExists('Disconnect');
+  }
+
+  /**
+   * Test the form in a pre-configured state.
+   */
+  public function testPreConfigured() {
+    $this->testKeyGeneration();
+    $this->drupalGet('/admin/config/ai/providers/amazeeio');
+    $this->assertSession()->buttonExists('Disconnect');
+  }
+
+  /**
+   * Verify disconnect works as expected.
+   */
+  public function testDisconnect() {
+    $this->testKeyGeneration();
+    $this->getSession()->getPage()->pressButton('Disconnect');
+    // We have to press disconnect twice, for the confirmation.
+    $this->getSession()->getPage()->pressButton('Disconnect');
+    $this->assertSession()->fieldExists('Email');
+
+    $this->drupalGet('/admin/config/ai/providers/amazeeio');
+
+    $config = \Drupal::configFactory()->get(AmazeeioAiConfigForm::CONFIG_NAME);
+    $this->assertEquals('', $config->get('host'));
+    $this->assertEquals('', $config->get('postgres_default_database'));
+    $this->assertEquals('', $config->get('postgres_host'));
+    $this->assertEquals('', $config->get('postgres_username'));
+
+    $keyStorage = \Drupal::entityTypeManager()->getStorage('key');
+    $aiKey = $keyStorage->load(AmazeeioAiConfigForm::API_KEY_NAME);
+    $dbKey = $keyStorage->load(AmazeeioAiConfigForm::VDB_PASSWORD_NAME);
+    $this->assertNull($aiKey);
+    $this->assertNull($dbKey);
+  }
+
+}
diff --git a/tests/src/Kernel/MockHttpClientTest.php b/tests/src/Kernel/MockHttpClientTest.php
new file mode 100644
index 0000000000000000000000000000000000000000..0b8faf3800f682779e10ea5eba0393e97569ee42
--- /dev/null
+++ b/tests/src/Kernel/MockHttpClientTest.php
@@ -0,0 +1,74 @@
+<?php
+
+namespace Drupal\Tests\ai_provider_amazeeio\Kernel;
+
+use Drupal\KernelTests\KernelTestBase;
+use Drupal\ai_provider_amazeeio_test\MockHttpClient;
+use GuzzleHttp\ClientInterface;
+use GuzzleHttp\Utils;
+
+/**
+ * Test class for the API mocking framework.
+ */
+class MockHttpClientTest extends KernelTestBase {
+  /**
+   * {@inheritdoc}
+   */
+  protected static $modules = ['ai_provider_amazeeio_test'];
+
+  /**
+   * The http client instance.
+   */
+  protected ClientInterface $client;
+
+  /**
+   * Retrieve the http client.
+   */
+  public function client() : MockHttpClient {
+    return $this->container->get('http_client');
+  }
+
+  /**
+   * Requests to hosts that don't contain 'amazee' should pass through.
+   */
+  public function testUnmockedRequests() : void {
+    $response = $this->client()->get('https://www.drupal.org');
+    $this->assertEquals(
+      expected: 200,
+      actual: $response->getStatusCode(),
+    );
+  }
+
+  /**
+   * Unmocked paths to 'amazee' domains throw an exception.
+   */
+  public function testUnmockedPath() : void {
+    $this->expectExceptionMessage('MockHttpClient: Unhandled request GET:https://www.amazee.io/auth/login');
+    $this->client()->get('https://www.amazee.io/auth/login');
+  }
+
+  /**
+   * Client exceptions in mocks are thrown and can be handled.
+   */
+  public function testMockException() : void {
+    $this->expectExceptionMessage('Missing authentication header');
+    $this->client()->get('https://www.amazee.io/mocked/request', [
+      'body' => Utils::jsonEncode(['message' => 'hello']),
+    ]);
+  }
+
+  /**
+   * A successful mocked request.
+   */
+  public function testMockSuccess() : void {
+    $result = $this->client()->get('https://www.amazee.io/mocked/request', [
+      'headers' => ['Authentication' => 'Bearer 1234'],
+      'body' => Utils::jsonEncode(['message' => 'hello']),
+    ]);
+    $this->assertEquals(200, $result->getStatusCode());
+    $this->assertEquals([
+      'uppercase' => 'HELLO',
+    ], Utils::jsonDecode($result->getBody()->getContents(), TRUE));
+  }
+
+}

