diff --git a/src/Plugin/Oauth2GrantManager.php b/src/Plugin/Oauth2GrantManager.php
index 1a8e0e1..418f418 100644
--- a/src/Plugin/Oauth2GrantManager.php
+++ b/src/Plugin/Oauth2GrantManager.php
@@ -113,7 +113,7 @@ class Oauth2GrantManager extends DefaultPluginManager implements Oauth2GrantMana
     AccessTokenRepositoryInterface $access_token_repository,
     RefreshTokenRepositoryInterface $refresh_token_repository,
     ConfigFactoryInterface $config_factory,
-    ResponseTypeInterface $response_type = NULL
+    ?ResponseTypeInterface $response_type = NULL
   ) {
     parent::__construct('Plugin/Oauth2Grant', $namespaces, $module_handler, 'Drupal\simple_oauth\Plugin\Oauth2GrantInterface', 'Drupal\simple_oauth\Annotation\Oauth2Grant');
 
@@ -133,7 +133,7 @@ class Oauth2GrantManager extends DefaultPluginManager implements Oauth2GrantMana
   /**
    * {@inheritdoc}
    */
-  public function getAuthorizationServer($grant_type, ConsumerInterface $client = NULL) {
+  public function getAuthorizationServer($grant_type, ?ConsumerInterface $client = NULL) {
     try {
       /** @var \Drupal\simple_oauth\Plugin\Oauth2GrantInterface $plugin */
       $plugin = $this->createInstance($grant_type);
diff --git a/src/Plugin/Oauth2GrantManagerInterface.php b/src/Plugin/Oauth2GrantManagerInterface.php
index ec925d7..ec97d80 100644
--- a/src/Plugin/Oauth2GrantManagerInterface.php
+++ b/src/Plugin/Oauth2GrantManagerInterface.php
@@ -20,6 +20,6 @@ interface Oauth2GrantManagerInterface {
    * @return \League\OAuth2\Server\AuthorizationServer
    *   The authorization server.
    */
-  public function getAuthorizationServer($grant_type, ConsumerInterface $client = NULL);
+  public function getAuthorizationServer($grant_type, ?ConsumerInterface $client = NULL);
 
 }
